docs: incorporate tutorial feedback (#354)
canonical/imagecraft: docs/tutorials/build-an-ubuntu-image.rst

.. meta::
    :description: Learn how to build a custom Ubuntu image with Imagecraft.

.. _tutorial-build-an-ubuntu-image:

Build an Ubuntu image
=====================

In this tutorial, we'll build a custom Ubuntu image for AMD64 machines. We'll work
through everything from the initial project setup to the image's first boot.

The tutorial takes about 25 minutes to complete. It doesn't require an intimate
understanding of disk images, but you'll need to be familiar with Linux paradigms and

> terminal usage.

Reviewer: What do you think about:

```suggestion
understanding of disk images, but you'll need to be familiar with Linux paradigms and
using the terminal.
```

"familiar with terminal usage" sounds a bit academic, like I'm studying how terminals are used by others.

Author: I like that much better. Thanks! https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

What you'll build
-----------------

> After installing the necessary tools, you'll start building a custom Ubuntu image from

Reviewer: > you'll start building

In the intro, it says "we'll build".

I think we should be as exact as possible in our usage of "we" and "you". In the Snapcraft tutorial, I tried to stick to tutor-like interlocutory patterns. I used "we" when describing something that multiple students could be doing at their own computers. I used "you" for things that affect the individual student personally (gaining experience, knowledge), or that can only happen on their machine, or requires their choice.

Author: Fixed. The rest of the doc aligns with the pattern you describe here. I must've messed it up here looking at the subsequent section title.

https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

the ground up.

> The tutorial guides you through the process of defining the image's structure and

Reviewer: ```suggestion
Then, we'll guide you through the process of defining the image's structure and
```

Author: https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

Edit: [changed my mind](https://github.com/canonical/imagecraft/pull/354/commits/f56a82068dd411937ca4513e172f62f970e63777)

content step by step. The image will be based on the suite of packages from Ubuntu 24.04

> LTS, with some additional software to cater it to the tutorial.

Reviewer: ```suggestion
LTS, with some additional software that caters it to the tutorial.
```

Author: https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

Install prerequisites
---------------------

To begin, let's install the Imagecraft snap. Open a terminal and run:

.. literalinclude:: code/build-an-ubuntu-image/task.yaml
    :language: bash
    :dedent: 2
    :start-at: snap install imagecraft --beta --classic
    :end-at: snap install imagecraft --beta --classic

Next, let's install Multipass, which will create the build environment when it comes

> time to package the image.

Reviewer: ```suggestion
time to package the image:
```

This is quite subtle, but I've come around on colons introducing the next code action. There are times in the prose when we need to pause and reflect on an action, or provide non-code instructions. Such paragraphs would end with regular points.

There can also be other box elements after a paragraph, and when skimming they could be confused for code.

The colon before a code block acts as a succinct signpost to say "yes, this paragraph is a new instruction, which requires you to run this in your terminal:"

Author: https://github.com/canonical/imagecraft/pull/354/commits/dea01f0b44a0ae858aad25b7c4f6dc3a5ff35651

.. literalinclude:: code/build-an-ubuntu-image/task.yaml
    :language: bash
    :dedent: 2
    :start-at: snap install multipass
    :end-at: snap install multipass

We'll run our image with QEMU. Install it with:

.. literalinclude:: code/build-an-ubuntu-image/task.yaml
    :language: bash
    :dedent: 2
    :start-at: sudo apt install qemu-system-x86
    :end-at: sudo apt install qemu-system-x86

Lastly, we'll need UEFI firmware to pass to QEMU. One of the most popular choices is
OVMF.

.. literalinclude:: code/build-an-ubuntu-image/task.yaml
    :language: bash
    :dedent: 2
    :start-at: sudo apt install ovmf
    :end-at: sudo apt install ovmf

Set up the project
------------------

We'll need a directory to hold the project. Create a directory wherever you like to keep

> your software projects with:

Reviewer: Could cut "with".

    :start-at: imagecraft init
    :end-at: imagecraft init

Open the generated project file in your preferred text editor. This is where most of the
tutorial will take place.

.. _tutorial-describe-the-image:

Describe the image
------------------

An image's project file starts with details like its name, version, and build

> environment. The ``init`` command set these keys to some generic values. Let's

Reviewer: ```suggestion
environment. The ``init`` command set these keys to generic values. Let's
```

Author: https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

> environment. The ``init`` command set these keys to some generic values. Let's

Reviewer: > `init` command

To cut down on inline code, we could say "when Imagecraft initialized the project".

Author: https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

We set the ``type`` key to the identifier for EFI system partitions in a GUID partition
table. The ``rootfs`` partition was generated with the identifier for Linux file

> systems. The identifiers themselves come from the UEFI specification--don't worry about

Reviewer: Technically, en dashes should be surrounded by spaces. Em dashes are the ones that touch their neighbours.

Author: Agh, good catch.

https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

    :emphasize-lines: 5, 6

With this entry, we mounted the EFI system partition to the /boot/efi/ directory in the
final image. We'll create this directory shortly.

.. _tutorial-set-up-the-root-file-system:

Set up the root file system
---------------------------

Because we're building the image on the bare base, its file system is currently an empty
directory. Let's start building it up with the ``parts`` key.

*Parts* are the means by which we source packages for and manipulate the files in the
image. They're the primary way we interact with the *overlay file system*, which is

> where we'll build up the image's content.

Reviewer: Did we discuss singular content? I'm not a great fan of it. It makes me think of web media.

Author: Most uses of "content" in the doc were actually redundant. I'm glad you made brought this to my attention :)

https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f

We'll create the file system with a part that uses the mmdebstrap plugin.

In the ``parts`` key, replace the template part with a new part named ``rootfs``,
defined as follows:

.. literalinclude:: code/build-an-ubuntu-image/imagecraft.yaml
    :language: yaml
    :start-at: parts:
    :end-at: noble

The ``mmdebstrap-suite`` key is specific to the mmdebstrap plugin and specifies the
package suite to bootstrap. To get the packages that are shipped with Ubuntu 24.04 LTS,
we set the suite to ``noble``.

The plugin removes the default sources configuration files, which limit us to the system
packages from the ``noble`` suite's ``main`` component. If we want to install anything
more than essential system packages, we'll need to add a new sources configuration file.

> We also still need to create the ``/boot/efi`` directory we mounted the ``efi``

Reviewer: > /boot/efi

Should we add a trailing slash?

Author: Indeed we should.

https://github.com/canonical/imagecraft/pull/354/commits/6300ad1c01e142bca78d753a3f1960bbf6eb0c9f
